# Overview
There is a queue at a festival for people to fill up their water bottles. Your task is to calculate the total time required for each person to fill up their water bottle.

### Requirement
You’ll need to create a function that takes two inputs:
1. An array of integers: which represents a queue of people, and each integer is the size of their water bottle in millilitres. For instance, the array of:
[400, 750, 1000]
represents a queue of three people, where the first person in the queue has a water bottle of 400ml,
the second person in the queue has a water bottle of 750ml, and the last person in the queue has a
water bottle of 1 litre.
2. An integer: which represents the number of taps at the festival available for people to use for filling up their water bottles.
This function should then return the total number of seconds that it takes for all people to have filled their water bottles.

### Assumptions
You must assume that as soon as one tap is free, the next person in the queue starts using that tap and that they cannot move to a different tap once they start filling their bottle. You must also assume that each tap flows at a rate of 100ml per second (e.g., a 1 litre bottle takes 10 seconds).


In [180]:
def calculate_time(queue, number_of_taps):
    
    # initializes a list that represent the working taps and contain the time required to fill the current bottle at each tap 
    # initializes a time counter for the complete task
    taps = [0]*number_of_taps
    time = 0
    
    # the algorithm goes through all bottles until none are left in the queue
    # the same logic applies for assigning a bottle to a tap irrespective of whether all taps are availble (like at the begining of the task) or only one of them is available (as the task is in progress)
    while queue:
        
        # finds the bottle that requires the least anmount of time to fill and the tap index for it
        fastest_to_fill = min(taps)
        tap_index = taps.index(fastest_to_fill)
        
        # deducts that time from all other bottles that are currently being filled by other taps 
        taps = [time_to_fill - fastest_to_fill for time_to_fill in taps]
        
        # adds that time to the total time counter of the task
        time+= fastest_to_fill
        
        # assigns the time to fill the next bottle in the queue to the tap that became available
        # this is also the part of the code that will prevent an infinite loop
        taps[tap_index] = queue.pop(0)/100
    
    # once no bottles are left in the queue, we find amongst those that are still being filled the one that would take the longest time and add it to the time counter to get the total time for the task 
    time+= max(taps)
    return time

    

##### Some test cases

In [181]:
calculate_time([500, 500, 500, 500, 500, 500], 3)

10.0

In [182]:
calculate_time([500, 500, 500, 500, 500, 500], 2)

15.0

In [183]:
calculate_time([100, 100, 100, 100, 100, 100], 4)

2.0

##### Same bottles in the queue but in differet order

In [184]:
calculate_time([400, 500, 300, 750, 250, 800], 3)

13.0

In [185]:
calculate_time([400, 750, 300, 500, 250, 800], 3)

14.5

In [197]:
calculate_time([400, 300, 800, 500, 250, 750], 3)

14.0

- As we can see from the above examples the order of the bottles in the queue matters and affects the overall time to complete the task

##### What if we tried to sort the bottles in the queue?

In [188]:
queue = [800, 750, 300, 500, 250, 400]
print(sorted(queue))
calculate_time(queue, 3)


[250, 300, 400, 500, 750, 800]


12.0

In [189]:
queue = [800, 750, 300, 500, 250, 400]
print(sorted(queue, reverse = True))
calculate_time(queue, 3)


[800, 750, 500, 400, 300, 250]


12.0

- It does seem to somewhat improve the time, however, as we can see in the next examples sorting the list isn't necessairly going to achieve the most efficient time  

In [186]:
calculate_time([800, 750, 500, 250, 300, 400], 3)

11.5

In [192]:
calculate_time([300, 750, 800, 500, 400, 250], 3)

11.5

- The most efficient time is a result of taps being switched off (i.e. all bottles are filled) as close as possible to one another. The closer we are to that scenario, the shorter the time gets  

### Bonus Points
Extra consideration will be given to candidates who complete the bonus points below:


#### Input validation
To stop someone using your function incorrectly, you should validate the inputs to the function and throw an error if they fail your validation.


In [143]:
# input validation function
def validate_input(queue, number_of_taps):
    
    # input validation starts with the number of taps, which has to be an integer number. an apropriate message will be displayed to the user in case it isn't  
    try:
         assert type(number_of_taps) == int
    except AssertionError:
        print(f"number of taps has to be an integer number, {number_of_taps} is not")
        return False

    # validation continues to ensure the number of taps is positive too. an apropriate message will be displayed to the user in case it isn't  
    try:
        assert number_of_taps > 0
    except AssertionError:
        print(f"number of taps has to be a positive number, {number_of_taps} is not")
        return False

    # validation moves to the queue input. Since the expected parameter here is a list, we need to ensure the user input matches the expected type
    try:
        assert type(queue) == list
    except AssertionError:
        print(f"the first input parameter must be of type list, {queue} is not")
        return False

    # once we established queue is a list, we will validate its content 
    for item in queue:

        # just like with validating the number of taps, the input for each bottle in the queue has to be an integer number 
        try:
            assert type(item) == int
        except AssertionError:
            print(f"all items in the queue must be of integer type {item} is not")
            return False

        # and it must also be positive
        try:
            assert item > 0
        except AssertionError:
            print(f"all items in the queue must be a positive number {item} is not")
            return False
    return True

def calculate_time1(queue, number_of_taps):
    
    # validation takes place and only valid input will get to be executed by the rest of the function
    if not(validate_input(queue, number_of_taps)):
        return
    
    taps = [0]*number_of_taps
    time = 0
    while queue:
        fastest_to_fill = min(taps)
        tap_index = taps.index(fastest_to_fill)
        taps = [time_to_fill - fastest_to_fill for time_to_fill in taps]
        time+= fastest_to_fill
        taps[tap_index] = queue.pop(0)/100
    time+= max(taps)
    return time

##### Input validation test cases

In [144]:
calculate_time2([400, 'a', 300, 750, 250, 800], 3)

all items in the queue must be of integer type a is not


In [145]:
calculate_time2([400, 500, -300, 750, 250, 800], 3)

all items in the queue must be a positive number -300 is not


In [146]:
calculate_time2([400, 500.7, 300, 750, 250, 800], 3)

all items in the queue must be of integer type 500.7 is not


In [147]:
calculate_time2([400, 500, 300, 750, 250, 800], '#')

number of taps has to be an integer number, # is not


In [148]:
calculate_time2([400, 500, 300, 750, 250, 800], 0)

number of taps has to be a positive number, 0 is not


In [149]:
calculate_time2([], 3)

0

#### Time to walk to tap
In the original challenge, we assumed that one person starts filling their water bottle as soon as the previous person finished. In reality this wouldn’t happen, since it takes time for each person to walk to the tap and open their bottle etc. You should write another function (which adapts your previous one slightly) which assumes that it takes a fixed amount of time (e.g., 2 seconds or 5 seconds) for each person to walk from the queue to the tap. (You may either assume that the initial people start at the tap, or they start at the queue and have to
walk to the tap. Either is fine.)


In [139]:
def calculate_time2(queue, number_of_taps):
    
    if not (validate_input(queue,number_of_taps)):
        return

    # initialize the fixed time it takes to walk to a tap (could be any given number of seconds, 5 seconds were used as an example)
    # it's also possible to have that as an input variable taken by the function, in which case we will have to validate it as numeric & positive (can be float)
    walk_to_tap_time = 5
        
    taps = [0]*number_of_taps
    time = 0
    while queue:
        fastest_to_fill = min(taps)
        tap_index = taps.index(fastest_to_fill)
        taps = [time_to_fill - fastest_to_fill for time_to_fill in taps]
        time+= fastest_to_fill
        
        # add walk to tap time to the time it takes to fill each bottle
        taps[tap_index] = queue.pop(0)/100 + walk_to_tap_time
        
    time+= max(taps)
    return time


In [140]:
calculate_time2([400, 500, 300, 750, 250, 800], 3)

23.0

#### Different flow rates of taps
In the original challenge, we assumed that each tap flows at a rate of 100ml per second. In reality, different taps might have more water pressure than others. You should write another function (which adapts your previous one slightly) to take account of the fact that different taps may flow at different rates (e.g., the first tap flows at 50ml per second while the second tap flows at 200ml per second). You should still assume that each person uses the first available tap.

In [275]:
# validation changes, both parameters are expected to be lists that contain the same type of input - positive integer numbers
# so we can pass both inputs in a list, and perform the same check for each of them
def validate_input2(input_lists):
 
    for input_list in input_lists:
        try:
            assert type(input_list) == list
        except AssertionError:
            print(f"both input parameters must be of type list, {input_list} is not")
            return False

        # it's important here to ensure the lists aren't empty 
        # there's no point for writing this function if there are no bottles in the queue or no taps to fill them
        try:
            assert len(input_list) > 0
        except AssertionError: 
            print(f"both input parameters cannot be empty. {input_list} is not an accepted input")
            return False
        
        for item in input_list:
            try:
                assert type(item) == int
            except AssertionError:
                print(f"input items in all lists must be of integer type {item} is not")
                return False
                
            try:
                assert item > 0
            except AssertionError:
                print(f"input items in all lists must be a positive number {item} is not")
                return False
    return True


# assuming different flow rates for each tap the function would now take a list of flow rates each one represents a tap   
def calculate_time3(queue, taps_flow_rate):
    
    if not (validate_input2([queue, taps_flow_rate])):
        return
    
    walk_to_tap_time = 5

    # the number of taps is inferred by the length of the flow rates list
    taps = [0]*len(taps_flow_rate)
    
    time = 0
    while queue:
        fastest_to_fill = min(taps)
        tap_index = taps.index(fastest_to_fill)
        taps = [time_to_fill - fastest_to_fill for time_to_fill in taps]
        time+= fastest_to_fill

        # account for the tap flow rate instead of the fixed rate of 100ml per second
        taps[tap_index] = queue.pop(0)/taps_flow_rate[tap_index] + walk_to_tap_time
        print(time, taps) 
    time+= max(taps)
    return time

##### Input validation test cases

In [56]:
calculate_time3([400, 500, 300, 750, 250, 800], [])

both input parameters cannot be empty. [] is not an accepted input


In [209]:
calculate_time3([400, 500, 300, 750, 250, 800], 9)

both input parameters must be of type list, 9 is not


In [210]:
calculate_time3([400, 500, -70, 750, 250, 800], [])

input items in all lists must be a positive number -70 is not


In [211]:
calculate_time3([400, 500, 170, 750, 250, 800], ['a','b'])

input items in all lists must be of integer type a is not


In [212]:
calculate_time3([400, 78.09, 300, 750, 250, 800], [4,5,6])

input items in all lists must be of integer type 78.09 is not


#### Faster taps, slower time
According to your function from bonus point 3 above, is it possible for your function to output a larger number (e.g., it takes longer to fill all the bottles) if you increase the flow rate of at least one of the taps (it takes less time to fill a bottle). If yes, find an example. If no, prove it

this is a case sceanrio that we've already seen earlier on

In [226]:
calculate_time3([400, 500, 300, 750, 250, 800], [100, 100, 100])

23.0

let's try to increase the flow rate of a tap and see what impact it has on the overall time...

In [227]:
calculate_time3([400, 500, 300, 750, 250, 800], [200, 100, 100])

23.0

intuitively one might think that when a tap fills more water per second, it will have an impact and improve the overall time... not necessarily. what if we increased the flow rate further?

In [239]:
calculate_time3([400, 500, 300, 750, 250, 800], [500, 100, 100])

23.0

the overall time still stays the same. maybe there's a tap that is more dominant for this task? what if we increase flow rate for a different tap?

In [231]:
calculate_time3([400, 500, 300, 750, 250, 800], [100, 100, 500])

23.0

In [274]:
calculate_time3([400, 500, 300, 750, 250, 800], [100, 125, 100])

20.5

increasing the flow rate of tap 3 isn't changing the overall time either, but increasing the flow rate of tap 2 ever so slightly (compared to the increases applied to the other two), seems to make a significant difference to the overall task time. let's see the taps in action- print statement added to show values stored in time and taps variables 

##### initial scenario all flow rates are the same

In [251]:
calculate_time3([400, 500, 300, 750, 250, 800], [100, 100, 100])

0 [9.0, 0, 0]
0 [9.0, 10.0, 0]
0 [9.0, 10.0, 8.0]
8.0 [1.0, 2.0, 12.5]
9.0 [7.5, 1.0, 11.5]
10.0 [6.5, 13.0, 10.5]


23.0

##### Increasing flow rate of tap 1 to 500ml/s

In [258]:
calculate_time3([400, 500, 300, 750, 250, 800], [500, 100, 100])

0 [5.8, 0, 0]
0 [5.8, 10.0, 0]
0 [5.8, 10.0, 8.0]
5.8 [6.5, 4.2, 2.2]
8.0 [4.3, 2.0, 7.5]
10.0 [2.3, 13.0, 5.5]


23.0

##### Increasing flow rate of tap 3 to 500ml/s

In [253]:
calculate_time3([400, 500, 300, 750, 250, 800], [100, 100, 500])

0 [9.0, 0, 0]
0 [9.0, 10.0, 0]
0 [9.0, 10.0, 5.6]
5.6 [3.4000000000000004, 4.4, 6.5]
9.0 [7.5, 1.0, 3.0999999999999996]
10.0 [6.5, 13.0, 2.0999999999999996]


23.0

ok, so we can see that when we increased flow rates of taps 1 & 3 they did fill bottles quicker, but it's the 2nd tap that ended up with the last bottle from the queue and that takes 13 seconds at a rate of 100ml/s (5sec walk + 8sec fill). if we increase the flow rate of one of those taps to something exteremly large, in a way that reduces fill time of any bottle unrealistically to no time, only then it will change the allocation of the last bottle and then we could achieve a faster time

##### Increasing flow rate of tap 1 to 5000ml/s

In [259]:
calculate_time3([400, 500, 300, 750, 250, 800], [5000, 100, 100])

0 [5.08, 0, 0]
0 [5.08, 10.0, 0]
0 [5.08, 10.0, 8.0]
5.08 [5.15, 4.92, 2.92]
8.0 [2.2300000000000004, 2.0, 7.5]
10.0 [0.23000000000000043, 13.0, 5.5]


23.0

##### Increasing flow rate of tap 1 to 500,000,000,000,000,000ml/s (!)

In [267]:
calculate_time3([400, 500, 300, 750, 250, 800], [500000000000000000, 100, 100])

0 [5.000000000000001, 0, 0]
0 [5.000000000000001, 10.0, 0]
0 [5.000000000000001, 10.0, 8.0]
5.000000000000001 [5.000000000000002, 4.999999999999999, 2.999999999999999]
8.0 [2.0000000000000027, 2.0, 7.5]
10.0 [2.6645352591003757e-15, 13.0, 5.5]


23.0

##### Increasing flow rate of tap 1 to 5,000,000,000,000,000,000ml/s (!!!)

In [266]:
calculate_time3([400, 500, 300, 750, 250, 800], [5000000000000000000, 100, 100])

0 [5.0, 0, 0]
0 [5.0, 10.0, 0]
0 [5.0, 10.0, 8.0]
5.0 [5.0, 5.0, 3.0]
8.0 [2.0, 2.0, 7.5]
10.0 [5.0, 0.0, 5.5]


15.5

it had to take that many zeros! and i don't even know the name for a number followed by 18 zeros

that was kinda fun, but now let's go back to our 2nd tap- the interesting one in this case...

In [270]:
calculate_time3([400, 500, 300, 750, 250, 800], [100, 125, 100])

0 [9.0, 0, 0]
0 [9.0, 9.0, 0]
0 [9.0, 9.0, 8.0]
8.0 [1.0, 1.0, 12.5]
9.0 [7.5, 0.0, 11.5]
9.0 [7.5, 11.4, 11.5]


20.5

increasing the flow by only an additional 25ml per second acheives a scenario, whereby compared to the original scenario, we have less idle time towards completion. <br><b>the closer idle time is to 0, the more productive each tap is and the faster the task will complete.</b> 

what happens if we increase the flow rate of tap no. 2 further? is there a scenario that will achieve a longer time at a higher flow rate?

In [271]:
calculate_time3([400, 500, 300, 750, 250, 800], [100, 200, 100])

0 [9.0, 0, 0]
0 [9.0, 7.5, 0]
0 [9.0, 7.5, 8.0]
7.5 [1.5, 8.75, 0.5]
8.0 [1.0, 8.25, 7.5]
9.0 [13.0, 7.25, 6.5]


22.0

In [272]:
calculate_time3([400, 500, 300, 750, 250, 800], [100, 500, 100])

0 [9.0, 0, 0]
0 [9.0, 6.0, 0]
0 [9.0, 6.0, 8.0]
6.0 [3.0, 6.5, 2.0]
8.0 [1.0, 4.5, 7.5]
9.0 [13.0, 3.5, 6.5]


22.0

- somewhat counter intuitive, but the answer is apparently <b>yes!</b> incrasing the flow rate for tap 2 from 125ml/s to 200ml/s or even 500ml/s results in an additional 1.5 seconds to the overall time
- as we can see at 125ml/s tap 2 with the highest capacity ends up as the one to fill the last bottle in the queue, and it's only idle for 0.1s while tap 1 is idle for 4 seconds
- by increasing the flow rate of tap 2 further to 200ml/s or 500ml/s, tap 2 fills bottles quicker, the allocation of bottles from the queue to available taps changes, and it's tap 1 that ends up with the last bottle from the queue while the time of tap 2 (our best worker) is wasted as it's idle for nearly 6s or 10s respectively, meaning the workload ditribution isn't the most efficient and doesn't maximize the productivity of our taps